processing/src/main/java/org/apache/druid/query/aggregation/post/ExpressionPostAggregator.java

```diff
@@ -100,14 +102,38 @@ private ExpressionPostAggregator(
     this.macroTable = macroTable;
     this.finalizers = finalizers;
 
-    this.parsed = Parser.parse(expression, macroTable);
-    this.dependentFields = ImmutableSet.copyOf(Parser.findRequiredBindings(parsed));
+    this.parsed = Suppliers.memoize(() -> Parser.parse(expression, macroTable));
+    this.dependentFields = Suppliers.memoize(() -> ImmutableSet.copyOf(Parser.findRequiredBindings(parsed.get())));
   }
 
+  private ExpressionPostAggregator(
```

Comment:
Could you make the other constructor just call this one, passing in the suppliers it is creating, so we don't have to duplicate all the other assignment logic?

Reply:
Well, I actually deliberately copied the code as Java force the constructor call to be the first statement. So to avoid code copying, I have to introduce one extra constructor, please check which one is preferred.